cpp/dolfinx/io/checkpointing.cpp

```diff
@@ -0,0 +1,354 @@
+  mesh::CellType cell_type;
+  basix::element::lagrange_variant lagrange_variant;
+
+  // Read attributes
+  {
+    adios2::Attribute<std::string> var_name
+        = io->InquireAttribute<std::string>("name");
+    adios2::Attribute<std::int32_t> var_dim
+        = io->InquireAttribute<std::int32_t>("dim");
+    adios2::Attribute<std::string> var_cell_type
+        = io->InquireAttribute<std::string>("cell_type");
+    adios2::Attribute<std::int32_t> var_degree
+        = io->InquireAttribute<std::int32_t>("degree");
+    adios2::Attribute<std::string> var_variant
+        = io->InquireAttribute<std::string>("lagrange_variant");
+
+    name = var_name.Data()[0];
+    dim = var_dim.Data()[0];
+    cell_type = mesh::to_type(var_cell_type.Data()[0]);
+    degree = var_degree.Data()[0];
+    lagrange_variant = string_to_variant[var_variant.Data()[0]];
+  }
+
+  // Scalar variables
+  std::uint64_t num_vertices_global;
+  std::uint64_t num_cells_global;
+  std::uint32_t num_dofs_per_cell;
+
+  // Read scalar variables
+  {
+    adios2::Variable<std::uint64_t> var_num_vertices
+        = io->InquireVariable<std::uint64_t>("num_vertices");
+    adios2::Variable<std::uint64_t> var_num_cells
+        = io->InquireVariable<std::uint64_t>("num_cells");
+    adios2::Variable<std::uint32_t> var_num_dofs_per_cell
+        = io->InquireVariable<std::uint32_t>("num_dofs_per_cell");
+
+    engine->Get(var_num_vertices, num_vertices_global);
+    engine->Get(var_num_cells, num_cells_global);
+    engine->Get(var_num_dofs_per_cell, num_dofs_per_cell);
+  }
+
+  // Compute local sizes, offsets
+  std::array<std::int64_t, 2> _local_range
+      = dolfinx::MPI::local_range(rank, num_vertices_global, size);
+
+  std::array<std::uint64_t, 2> local_range{(std::uint64_t)_local_range[0],
+                                           (std::uint64_t)_local_range[1]};
+  std::uint64_t num_vertices_local = local_range[1] - local_range[0];
+
+  std::array<std::int64_t, 2> _cell_range
+      = dolfinx::MPI::local_range(rank, num_cells_global, size);
+
+  std::array<std::uint64_t, 2> cell_range{(std::uint64_t)_cell_range[0],
+                                          (std::uint64_t)_cell_range[1]};
+  std::uint64_t num_cells_local = cell_range[1] - cell_range[0];
+
+  std::vector<int64_t> input_global_indices(num_vertices_local);
+  std::vector<T> x(num_vertices_local * 3);
+  std::vector<int64_t> array(num_cells_local * num_dofs_per_cell);
+  std::vector<int32_t> offsets(num_cells_local + 1);
+
+  {
+    adios2::Variable<std::int64_t> var_input_global_indices
+        = io->InquireVariable<std::int64_t>("input_global_indices");
+
+    adios2::Variable<T> var_x = io->InquireVariable<T>("x");
+
+    adios2::Variable<std::int64_t> var_topology_array
+        = io->InquireVariable<std::int64_t>("topology_array");
+
+    adios2::Variable<std::int32_t> var_topology_offsets
+        = io->InquireVariable<std::int32_t>("topology_offsets");
+
+    if (var_input_global_indices)
```

Comment:
If the variables don't exist you cannot return meaningful output. Raise an exception immediately after inquiring.